src/psyclone/dynamo0p3.py

```diff
@@ -7114,6 +7114,36 @@ def __init__(self, loop_type="", **kwargs):
         self._upper_bound_name = None
         self._upper_bound_halo_depth = None
 
+    def lower_to_language_level(self):
+        '''In-place replacement of DSL or high-level concepts into generic
+        PSyIR constructs. This function replaces a DynLoop with a PSyLoop
+        and inserts the loop boundaries into the new PSyLoop, or removes
+        the loop node in case of a domain kernel. Once TODO #1731 is done
+        (which should fix the loop boundaries, which atm rely on index of
+        the loop in the schedule, i.e. can change when transformations are
+        applied), this function can likely be removed.
+
+        '''
+        super().lower_to_language_level()
+        if self._loop_type != "null":
+            # Not a domain loop, i.e. there is a real loop
+            # We need to copy the expressions, since the original ones are
+            # attached to the original loop.
+            psy_loop = PSyLoop.create(self._variable,
+                                      self.start_expr.copy(),
+                                      self.stop_expr.copy(),
+                                      self.step_expr.copy(),
+                                      self.loop_body.pop_all_children())
+            self.replace_with(psy_loop)
+        else:
+            # Domain loop, i.e. no need for a loop at all. Remove the loop
+            # node (self), and insert its children directly
+            pos = self.position
+            parent = self.parent
+            self.detach()
+            for child in self.loop_body.pop_all_children():
```

Comment:
I think you need to reverse this list in order to preserve the order of the body statements.